src/pixie.c

```diff
@@ -18,20 +18,138 @@ void pixie_format(const unsigned char *in, unsigned len, char *outbuf) {
 void pixie_attack(void) {
 	struct pixie *p = &pixie;
 	int dh_small = get_dh_small();
+	int ret_val;
 
 	if(p->do_pixie) {
 		char cmd[4096];
 		snprintf(cmd, sizeof cmd,
 		"pixiewps -e %s -s %s -z %s -a %s -n %s %s %s",
 		p->pke, p->ehash1, p->ehash2, p->authkey, p->enonce,
 		dh_small ? "-S" : "-r" , dh_small ? "" : p->pkr);
-		printf("executing %s\n", cmd);
-		exit(system(cmd));
+		if (p->do_pixie == 2) {
+			ret_val = pixie_wrapper(cmd);
+			if (ret_val != EXIT_SUCCESS) {
+				/* <sys/wait.h> WEXITSTATUS() or div 256 */
+				exit(ret_val);
+			}
+		}
+		else {
+			printf("executing %s\n", cmd);
+			exit(system(cmd));
+		}
 	}
+}
+
+void pixie_free() {
 	PIXIE_FREE(authkey);
 	PIXIE_FREE(pkr);
 	PIXIE_FREE(pke);
 	PIXIE_FREE(enonce);
 	PIXIE_FREE(ehash1);
 	PIXIE_FREE(ehash2);
+	PIXIE_FREE(wrapper);
+	PIXIE_FREE(pin);
+}
+
+/**
+ * Call a wrapper to run pixiewps.
+ * 
+ * The wrapper could be implemented in any language which needs to follow the following conventions:
+ * 1) arguments are passed precisely as for pixiewps
+ * 2) on failure, non-zero exit status is returned
+ * 3) on success, exactly one line is printed to stdout, containing the found pin and a trailing newline character,
+ *    in case of null pin, only a newline character '\n' would be returned and exit status 0 is returned.
+ * 
+ * @params char* cmd The command line to call pixiewps with the arguments
+ * 
+ * @return int
+ */
+int pixie_wrapper(char *cmd) {
+	char *sh = NULL;
+	int ret_val = EXIT_FAILURE;
+
+	sh = (char*) malloc(strlen(pixie.wrapper) + strlen(cmd));
+	if (sh) {
+		/* Call the wrapper with the arguments passed */
+		strcpy(sh, pixie.wrapper);
```

Comment:
why not use snprintf() instead of the ugly mix of strcpy and strcat

Reply:
I based on the example at http://www.cplusplus.com/reference/cstring/strcat/ . Okay, I'll to use snprintf(). Thank!